File: 0085-maximal-rectangle/0085-maximal-rectangle.cpp
class Solution {
public:
    int returnMaxArea(vector<int>& heights) {


          stack<pair<int,int>> stackLeftToRight; 
        stack<pair<int,int>> stackRightToLeft; 
        int n = heights.size();
        vector<int> smallerElementIndexOnLeft(n,-1);
        vector<int> smallerElementIndexOnRight(n,n);
        for(int i=0;i<n;i++) {
            if(stackLeftToRight.empty()) {
                stackLeftToRight.push({heights[i],i});
            } else {
                while(stackLeftToRight.empty()==false && stackLeftToRight.top().first>=heights[i]) {
                    stackLeftToRight.pop();
                }
                if(stackLeftToRight.empty()) {
                    stackLeftToRight.push({heights[i],i});
                } else {
                    smallerElementIndexOnLeft[i]=stackLeftToRight.top().second;
                    stackLeftToRight.push({heights[i],i});


                }
            }
        }

        for(int i=n-1;i>=0;i--) {
            if(stackRightToLeft.empty()) {
                stackRightToLeft.push({heights[i],i});
            } else {
                while(stackRightToLeft.empty()==false && stackRightToLeft.top().first>=heights[i]) {
                    stackRightToLeft.pop();
                } 
               if(stackRightToLeft.empty()) {
                    stackRightToLeft.push({heights[i],i});
                } else {
                    smallerElementIndexOnRight[i]=stackRightToLeft.top().second;
                    stackRightToLeft.push({heights[i],i});


                }
            }
        }

    int ans=INT_MIN;
    for(int i=0;i<n;i++) {
    ans = max(ans,(smallerElementIndexOnRight[i]-smallerElementIndexOnLeft[i]-1)*heights[i]);   
    }
        return ans;
    }

    
    
    
    int maximalRectangle(vector<vector<char>>& matrix) {
        vector<vector<int>> heights(matrix.size(),vector<int>(matrix[0].size(),0));
        for(int i=0;i<matrix[0].size();i++) {
            if(matrix[0][i]=='1') heights[0][i]=1;
        }
        int ans=0;

        for(int r=1;r<matrix.size();r++) {
            for(int c=0;c<matrix[0].size();c++) {
                if (matrix[r][c]=='1') {

                 heights[r][c]=1+heights[r-1][c];
            } else {
                heights[r][c] = 0 ;
            }
        }

        }

        cout<<"height matrix is "<<endl;
        for(auto row:heights) {
            for(auto col:row) cout<<col<<" ";
            cout<<endl;
        }


        for(int i=0;i<matrix.size();i++) ans=max(ans,returnMaxArea(heights[i]));
return ans;
    }

};

File: 0085-maximal-rectangle/0085-maximal-rectangle.py
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        matrix = [list(map(int, row)) for row in matrix]

        n = len(matrix[0])
        heights = [0] * n
        maxAnswer = 0
        for m in range(len(matrix)):
            for j in range(n):
                if matrix[m][j] == 1:
                    heights[j] += 1
                else:
                    heights[j] = 0
            NSL = [-1]*n
            NSR = [n]*n
            NSLStack = []
            NSRStack = []
            #First- next smallest neighbour to right (monotonically increasing stack)
            for index in range(n-1,-1,-1):
                while NSRStack and heights[NSRStack[-1]]>=heights[index]:
                    NSRStack.pop()
                if len(NSRStack)>0:
                    NSR[index] = NSRStack[-1]
                NSRStack.append(index)


            for index in range(0,n,+1):
                while NSLStack and heights[NSLStack[-1]]>=heights[index]:
                    NSLStack.pop()
                if len(NSLStack)>0:
                    NSL[index] = NSLStack[-1]
                NSLStack.append(index)
            
            for index,number in enumerate(heights):
                maxAnswer = max(maxAnswer,int(number)*(NSR[index]-NSL[index]-1))

            
        return maxAnswer